### PR TITLE
Data gatherers can fetch data from local file

### PR DESCRIPTION
**Pull request by wwwil**
Possible solution for #78, currently only implemented for the AKS data gatherer.

### CODE REVIEW

**Comment by wwwil** [comment]
@j-fuentes @charlieegan3 What do you think of this approach?

By trying to run Preflight like this I was able to find and fix #81, so I think it's worth doing something to make it easier to just run locally with fake data. Not sure if this is the best way though.

**Comment by j-fuentes** [comment]
I'd rather bypass the whole data-gathering process:

- add a command-line option like `--bypass-data-gathering-with-file` that points to a json file containing the raw input for the Rego evaluation. That raw input is:
```
{
  "data-gatherer ID 1": {... data-gatherer 1 data ...},
  "data-gatherer ID 2": {... data-gatherer 2 data ...},
  ...
}
```

- if that option is present, skip the execution of the data gatherers, and check that the enabled data-gatherers that are required by packages are present in the file.

- then change [the part that builds the input for the Rego evaluation](https://github.com/jetstack/preflight/blob/baa0bc73fb4d63761082fbd5494784c8c3ee7f6e/cmd/check.go#L345) so it uses the data from the JSON.

- Additionally, we could make schema validation on the data from the JSON, but that would require more work on the `DataGatherer` interface to provide methods for data validation.

**Comment by charlieegan3** [comment]
I think if there's a way of doing this without making changes to individual gatherers that'd be preferred.

**Comment by wwwil** [comment]
Thanks for the feedback!

I'd agree having to do this for every data gatherer is not ideal so I'm going to look for a more general approach.

I went for loading the data from the data gatherers as it maximises how 'end-to-end' the test is and keeps as much in common with a 'normal' run as possible. Data gatherers also use the relevant `struct` for data so this provides some validation of what's being loaded from file too.

I'd prefer to be able to bypass individual data gatherers. I'd also prefer to have each data gatherer's data in separate files. I think this will be easier to work with and is more like the format we would get it in anyway, for example from `intermediate` type output.

**Comment by wwwil** [comment]
@j-fuentes @charlieegan3 I've updated the implementation.

This now adds a `local` data gatherer that to read a local file. This currently has no validation on the data it reads, it is just assumed to be JSON and passed direct to OPA for evaluation.

If any of the data gatherer's config specify a `data-path` then while creating data gatherers a `local` data gatherer will be created which 'impersonates' that gatherer.

File: docs/configuration.md [hunk 1]
```diff
@@ -41,6 +41,7 @@ The following data gatherers are available:
 - [Google Kubernetes Engine](docs/datagatherers/gke.md)
 - [Amazon Elastic Kubernetes Service](docs/datagatherers/eks.md)
 - [Microsoft Azure Kubernetes Service](docs/datagatherers/aks.md)
+- [Local](docs/datagatheres/local.md)
```
Comment (charlieegan3) [inline docs/configuration.md]: Todo?

File: examples/fake.preflight.yaml [hunk 2]
```diff
@@ -0,0 +1,21 @@
+cluster-name: my-fake-cluster
```
Comment (charlieegan3) [inline examples/fake.preflight.yaml]: should this file be called `local.preflight.yaml`?

File: cmd/check.go [hunk 3]
```diff
@@ -130,6 +131,19 @@ func check() {
...
 			// TODO: create gatherer from config in a more clever way. We need to read gatherer config from here and its schema depends on the gatherer itself.
 			var dg datagatherer.DataGatherer
+			dataGathererConfig, ok := config.(map[string]interface{})
+			if !ok {
+				log.Fatalf("Cannot parse %s data gatherer config.", name)
+			}
+			// Check if this data gatherer's config specifies a data-path.
+			// If it does create a LocalDataGatherer to load this data but keep
+			// the name of the data gatherer it is impersonating so it can
+			// provide fake data.
```
Comment (charlieegan3) [inline cmd/check.go]:
```suggestion
			// provide local data.
```

File: examples/fake.preflight.yaml [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+data-gatherers:
+  aks:
+    # Fake data for AKS.
```
Comment (charlieegan3) [inline examples/fake.preflight.yaml]:
```suggestion
    # Local data for AKS.
```

File: examples/fake.preflight.yaml [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    data-path: ./examples/data/aks.json
+  gke:
+    # Fake data for GKE.
```
Comment (charlieegan3) [inline examples/fake.preflight.yaml]:
```suggestion
    # Local data for GKE.
```

**Comment by charlieegan3** [comment]
Closing in favor of #80 (since I don't have permissions from Wil for his branch)

**Pull request closed without merging**